ext/symengine/ruby_matrix.c

```diff
@@ -0,0 +1,709 @@
+    cbasic_c = NUM2ULONG(c_);
+
+    cresult = dense_matrix_new();
+
+    dense_matrix_submatrix(cresult, this, cbasic_r1, cbasic_c1, cbasic_r2,
+                           cbasic_c2, cbasic_r, cbasic_c);
+    result = Data_Wrap_Struct(c_dense_matrix, NULL, dense_matrix_free, cresult);
+    return result;
+}
+
+VALUE cmatrix_dense_rows(VALUE self)
+{
+    CDenseMatrix *this;
+    Data_Get_Struct(self, CDenseMatrix, this);
+    return ULONG2NUM(dense_matrix_rows(this));
+}
+
+VALUE cmatrix_dense_cols(VALUE self)
+{
+    CDenseMatrix *this;
+    Data_Get_Struct(self, CDenseMatrix, this);
+    return ULONG2NUM(dense_matrix_cols(this));
+}
+
+VALUE cmatrix_dense_add(VALUE self, VALUE operand)
+{
+    CDenseMatrix *this;
+    Data_Get_Struct(self, CDenseMatrix, this);
+
+    CDenseMatrix *cresult;
+    VALUE result;
+
+    cresult = dense_matrix_new();
+
+    char *s = rb_obj_classname(operand);
+
+    if (strcmp(s, "SymEngine::DenseMatrix") == 0) {
+        // Matrix Addition
+        CDenseMatrix *coperand;
+        Data_Get_Struct(operand, CDenseMatrix, coperand);
+
+        dense_matrix_add_matrix(cresult, this, coperand);
+
+        result = Data_Wrap_Struct(c_dense_matrix, NULL, dense_matrix_free,
+                                  cresult);
+
+    } else {
+        // Scalar Addition
+        basic_struct *coperand = basic_new_heap();
+        sympify(operand, coperand);
+
+        dense_matrix_add_scalar(cresult, this, coperand);
+        result = Data_Wrap_Struct(c_dense_matrix, NULL, dense_matrix_free,
+                                  cresult);
+        basic_free_heap(coperand);
+    }
+
+    return result;
+}
+
+VALUE cmatrix_dense_mul(VALUE self, VALUE operand)
+{
+    CDenseMatrix *this;
+    Data_Get_Struct(self, CDenseMatrix, this);
+
+    CDenseMatrix *cresult;
+    VALUE result;
+
+    cresult = dense_matrix_new();
+
+    char *s = rb_obj_classname(operand);
+
+    if (strcmp(s, "SymEngine::DenseMatrix") == 0) {
+        // Matrix Multiplication
+        CDenseMatrix *coperand;
+        Data_Get_Struct(operand, CDenseMatrix, coperand);
+
+        dense_matrix_mul_matrix(cresult, this, coperand);
+
+        result = Data_Wrap_Struct(c_dense_matrix, NULL, dense_matrix_free,
+                                  cresult);
+    } else {
+        // Scalar Multiplication
+        basic_struct *coperand = basic_new_heap();
+        sympify(operand, coperand);
+
+        dense_matrix_mul_scalar(cresult, this, coperand);
+        result = Data_Wrap_Struct(c_dense_matrix, NULL, dense_matrix_free,
+                                  cresult);
+        basic_free_heap(coperand);
+    }
+
+    return result;
+}
+
+VALUE cmatrix_dense_eq(VALUE self, VALUE operand)
+{
+    CDenseMatrix *this;
+    Data_Get_Struct(self, CDenseMatrix, this);
+
+    CDenseMatrix *cresult;
+    VALUE result;
+
+    cresult = dense_matrix_new();
+
+    char *s = rb_obj_classname(operand);
+
+    if (strcmp(s, "SymEngine::DenseMatrix") == 0) {
```

Comment:
You can use the `sympify_matrix` function I mentioned so that a SymEngine::DenseMatrix can be equal to a Ruby array.